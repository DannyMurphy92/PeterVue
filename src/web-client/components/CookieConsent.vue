<template>
  <div v-if="showConsentBanner" class="cookie-consent_wrapper">
    <span v-external-click="onExternalClick"
      >This website uses cookies as described in our <router-link :to="privacyPolicy">cookie policy</router-link>, to
      see what cookies we use and to set your own preferences, click here. Otherwise, by clicking on or navigating this
      site, you accept our use of cookies.
    </span>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import {hasConsentCookie, setConsentCookie} from '@/utils/cookieConsent'
import { Page } from '@/router/Page'

export default Vue.extend({
  name: 'CookieConsent',
  data() {
    return { showConsentBanner: true, privacyPolicy: { name: Page.PrivacyPolicy, hash: '#cookie-policy' } }
  },
  created() {
    if (hasConsentCookie()) {
      this.loadGa()
    }
  },
  methods: {
    onExternalClick(): void {
      setConsentCookie()
      this.loadGa()
    },
    loadGa(): void {
      this.showConsentBanner = false
      window.gtag('js', new Date())

      window.gtag('config', window.gaId)
      window.gtag('pageview')
    },
  },
})
</script>
<style lang="scss">
.cookie-consent {
  &_wrapper {
    bottom: 8px;
    position: fixed;
    background: #fff;
    border: 1px solid #000;
    border-radius: 2px;
    padding: 2px;
    margin: 0 20px;
  }
}
</style>
