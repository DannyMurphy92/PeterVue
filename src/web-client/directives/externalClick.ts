import { VNode, DirectiveOptions, VNodeDirective } from 'vue'

const directive: DirectiveOptions = {
  bind(element: any, binding: VNodeDirective, vnode: VNode) {
    element.onClickExternal = (event: Event) => {
      if (!(element === (event.target as HTMLElement) || element.contains(event.target as HTMLElement))) {
        if (typeof binding.value !== 'function') {
          console.warn('v-external-click directive used without a handler') // tslint:disable-line no-console
          return
        }
        binding.value.call(vnode, event)
      }
    }
    document.addEventListener('click', element.onClickExternal)
  },
  unbind(element: any) {
    document.removeEventListener('click', element.onClickExternal)
    delete element.onClickExternal
  },
}

export default directive
