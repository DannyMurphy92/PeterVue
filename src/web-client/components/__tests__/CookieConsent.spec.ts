import CookieConsent from '../CookieConsent.vue'
import { Wrapper, shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import VueRouter from 'vue-router'
import externalClick from '@/directives/externalClick'

Vue.use(VueRouter)
Vue.directive('external-click', externalClick)

let setConsentCookieSpy: jest.SpyInstance
let hasConsentCookieSpy: jest.SpyInstance
let hasConsentCookie = false

jest.mock('@/utils/cookieConsent', () => {
  setConsentCookieSpy = jest.fn()
  hasConsentCookieSpy = jest.fn(() => hasConsentCookie)
  return {
    hasConsentCookie: hasConsentCookieSpy,
    setConsentCookie:setConsentCookieSpy,
  }
})

describe('CookieConsent.vue', () => {
  let wrapper: Wrapper<any>
  const gtagSpy = jest.fn()

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('GIVEN initial state', () => {
    beforeEach(() => {
      window.gtag = gtagSpy
      wrapper = shallowMount(CookieConsent)
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('THEN renders the consent banner', () => {
      expect(wrapper.find('.cookie-consent_wrapper').exists()).toBe(true)
    })

    describe('WHEN click outside banner', () => {
      beforeEach(() => {
        document.dispatchEvent(new Event('click'))
      })

      it('THEN does not render the banner', () => {
        expect(wrapper.find('.cookie-consent_wrapper').exists()).toBe(false)
      })

      it('THEN sets up google analytics', () => {
        expect(gtagSpy).toBeCalledTimes(3)
        expect(gtagSpy).toBeCalledWith('config', 'TEST-GA-ID')
        expect(gtagSpy).toBeCalledWith('pageview')
      })

      it('THEN sets consent cookie', () => {
        expect(setConsentCookieSpy).toBeCalledTimes(1)
      })
    })
  })

  describe('WHEN has cookie consent cookie set', () => {
    beforeEach(() => {
      hasConsentCookie = true
    })

    describe('WHEN first mounted', () => {
      beforeEach(async () => {
        wrapper = shallowMount(CookieConsent)
        wrapper.vm.$forceUpdate()
      })

      it('sdfsf', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      it('THEN does not render the banner', () => {
        expect(wrapper.find('.cookie-consent_wrapper').exists()).toBe(false)
      })

      it('THEN does not set consent cookies', () => {
        expect(setConsentCookieSpy).not.toBeCalled()
      })
    })
  })
})
