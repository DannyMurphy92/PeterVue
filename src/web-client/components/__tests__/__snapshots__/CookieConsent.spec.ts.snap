// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieConsent.vue WHEN has cookie consent cookie set WHEN first mounted sdfsf 1`] = `
<div
  class="cookie-consent_wrapper"
>
  <span>
    This website uses cookies as described in our 
    <a
      class=""
      href="/privacy#cookie-policy"
    >
      cookie policy
    </a>
    , to
    see what cookies we use and to set your own preferences, click here. Otherwise, by clicking on or navigating this
    site, you accept our use of cookies.
  
  </span>
</div>
`;
